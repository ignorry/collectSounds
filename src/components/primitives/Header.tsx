import React, { useRef } from "react";
import styled from "styled-components";

import Back from "./Back";

const Wrapper = styled.div`
  position: absolute;
  width: 100%;
  background: ${ ({ theme }) => theme.colors.bgSecondary };
  padding: ${ ({ theme }) => `${ theme.gaps.small }rem ${ theme.gaps.small }rem`};
  z-index: ${ ({ theme }) => theme.order.header };
`;

const Collumn = styled.div`
  max-width: ${ ({ theme }) => `${ theme.contentWidth }rem` };
  margin: auto;
`;

const Fix = styled.div<{ height: number }>`
  height: ${ props => `${ props.height }px` };
`;

/**
 * @typedef Props
 * @prop {Function} [callback] - click callback
 */
export type Props = {
  callback?: Function,
}

const Link: React.FC<Props> = ( props: Props ) => {
  const wrapperRef: React.MutableRefObject<undefined | HTMLDivElement> = useRef();
  
  return (
    <>
      <Wrapper ref={ wrapperRef }>
        <Collumn>
          <Back
            callback={ props.callback }
          />
        </Collumn>
      </Wrapper>
      <Fix height={ wrapperRef.current ? wrapperRef.current.getBoundingClientRect().height : 0 }>
      </Fix>
    </>
  )
};

export default Link;