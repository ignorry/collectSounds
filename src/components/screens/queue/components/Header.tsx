import React, { useEffect, useRef } from "react";
import { useIntl } from "react-intl";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import { getQueueLength } from "../../../../lib/redux/getQueueLength";
import { cleanQueue } from "../../../../redux/slices/queue";

import Button from "../../../primitives/Button";
import Label from "../../../primitives/Label";
import ToggleButton from "../../../primitives/ToggleButton";

const Wrapper = styled.div`
  position: absolute;
  width: 100%;
  background: ${ ({ theme }) => theme.colors.bgSecondary };
  padding: ${ ({ theme }) => `${ theme.gaps.small }rem ${ theme.gaps.small }rem`};
  z-index: ${ ({ theme }) => theme.order.header };
`;

const Collumn = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  max-width: ${ ({ theme }) => `${ theme.contentWidth }rem` };
  margin: auto;
`;

const Fix = styled.div<{ height: number }>`
  height: ${ ({ theme }) => `${ theme.font.primary.size + theme.gaps.small*2 }rem` };
`;

const Link: React.FC = () => {
  const wrapperRef: React.MutableRefObject<undefined | HTMLDivElement> = useRef();
  const intl = useIntl();
  const dispatch = useDispatch();

  console.log( 'wrapper', [wrapperRef.current] );
  
  return (
    <>
      <Wrapper ref={ wrapperRef }>
        <Collumn>
          <Label text={`${ intl.formatMessage({ id: 'length' }) }: ${ dispatch( getQueueLength() as any ) }`} secondary/>

          <ToggleButton
            text={ intl.formatMessage({ id: 'clearQueue' }) }
            callback={ () => dispatch( cleanQueue() ) }
            active
          />
        </Collumn>
      </Wrapper>
      <Fix height={ wrapperRef.current ? wrapperRef.current.getBoundingClientRect().height : 0 }>
      </Fix>
    </>
  )
};

export default Link;