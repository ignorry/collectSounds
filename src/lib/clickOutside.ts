/**
 * Listen to clicks outside of the element
 * @param {HTMLElement} elem
 * @param {Function} callback 
 * @returns {void}
 */
export const clickOutside = ( elem: HTMLElement, callback: Function ): void =>
  document.addEventListener( 'click', (e) => {
    const target: Node = e.target as Node;
    
    if ( target !== elem && ! elem.contains( target ) )
      callback();
  });