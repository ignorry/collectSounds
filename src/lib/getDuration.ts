/**
 * get duration from ms in readable format (h:mm:ss)
 * @param {number} duration 
 * @returns {string}
 */
export const getDurationFromMs = ( duration: number ): string =>
  `${ duration/3600000 > 1 ? `${ Math.floor( duration/3600000 ) }:` : ''
  }${ duration/60000 > 1 ? `${ Math.floor( duration%3600000/60000 ) }:` : ''
  }${ `${ Math.floor( duration%60000/1000 ) }` }`