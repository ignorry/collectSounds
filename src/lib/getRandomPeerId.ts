/**
 * generates 64 characters id starting with 'collectS'
 * @returns { string }
 */
export const getRandomPeerId = () => {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  for ( let i = 0; i < 56; i++ ) {
    result += characters.charAt( Math.floor( Math.random() * characters.length ) );
  }

  return 'collectS' + result;
}
