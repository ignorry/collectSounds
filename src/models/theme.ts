export enum ThemeEnum  {
  light = 'light',
  dark = 'dark'
}

// all sizes in rem
export type Theme = {
  iconSize: number,
  rangeSize: number,
  spinnerSize: number,
  contentWidth: number,
  colors: {
    main: string,
    bg?: string,
    bgSecondary?: string,
    font?: string,
    darkBg?: string,
  },
  decorative: { // in px
    borderRadius: number;
    padding: number;
  },
  gaps: {
    smallest: number,
    small: number,
    big: number,
  },
  thumbnail: {
    width: number,
    height: number,
    infoColor: string,
    infoWidth: 4,
  },
  error: {
    backgroundColor: string,
    width: number,
  },
  media: {
    small: string,
  },
  font: {
    primary: {
      size: number,
      opacity: number,
    },
    secondary: {
      size: number,
      opacity: number,
    },
  },
  durations: {
    ms300: number,
  },
  order: {
    footer: number,
    modal: number,
  },
}