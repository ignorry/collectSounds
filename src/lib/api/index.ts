import { 
  fetchVideo,
  fetchPlaylist,
  fetchPlaylistItems,
  fetchChannel,
  fetchSearch
} from "./fetch";
import {
  parseVideo,
  parsePlaylist,
  parseChannel
} from "./parseApiResponse";

import { Video, Playlist, Channel, Content } from "../../models/content";

/**
 * fetch video by id
 * @param {string} id - video id
 * @returns {Promise<Array<Video>>} 
 */
export const getVideoById = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchVideo({
    part: 'snippet,contentDetails',
    id
  });

  return parseVideo( res );
}

/**
 * fetch playlist by id
 * @param {string} id - playlist id
 * @returns {Promise<Array<Playlist>>}
 */
export const getPlaylistById = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    id
  });

  return parsePlaylist( res );
}

/**
 * fetch channel by id
 * @param {string} id - channel id
 * @returns {Promise<Array<Channel>>}
 */
export const getChannelById = async ( id: string ): Promise<Array<Channel>> => {
  const res = await fetchChannel({
    part: 'snippet',
    id
  });

  return parseChannel( res );
}

/**
 * fetch playlist's videos
 * @param {string} id - playlist id
 * @returns {Promise<Array<Video>>}
 */
export const getPlaylistVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchPlaylistItems({
    part: 'id',
    playlistId: id,
  });

  const ids: Array<string> = res.items.map( item => item.id );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's videos
 * @param {string} id - channel id
 * @returns {Promise<Array<Video>>}
 */
export const getChannelVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchSearch({
    part: 'id',
    channelId: id,
    order: 'date',
  })

  const ids: Array<string> = res.items.filter( item => item.id.kind === 'video' ).map( item => item.id.videoId );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's playlists
 * @param {string} id - channel id
 * @returns {Promise<Array<Playlist>>}
 */
export const getChannelPlaylists = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    channelId: id,
  });

  return parsePlaylist( res )
}

/**
 * @typedef SearchFilters
 * @prop {string} order - specifies the method that will be used to order resources in the API response
 * @prop {string} type - restricts a search query to only retrieve a particular type of resource
 * @prop {string} videoDuration - filters video search results based on their duration
 * @prop {string} saveSearch - indicates whether the search results should include restricted content as well as standard content
 */
export type SearchFilters = {
  order?: 'date' | 'rating' | 'relevance' | 'title' | 'videoCount' | 'viewCount';
  type?: 'video' | 'playlist' | 'channel',
  videoDuration?: 'any' | 'long' | 'medium' | 'short';
  saveSearch?: 'moderate' | 'none' | 'strict'
}

/**
 * fetch search results with specified filters
 * @param {SearchFilters} filters
 * @returns {Promise<Array<Content>>}
 */
export const getSearch = async ( filters: SearchFilters ): Promise<Array<Content>> => {
  const res = await fetchSearch({...filters,
    part: 'id',
    maxResults: 10,
  });

  const videoIds: Array<string> = [];
  const playlistIds: Array<string> = [];
  const channelIds: Array<string> = [];
  const ids: Array<string> = res.items.map( item => {
    if ( item.id.kind === 'video' ) {
      videoIds.push( item.id.videoId );
      return item.id.videoId;
    } else if ( item.id.kind === 'playlist' ) {
      playlistIds.push( item.id.playlistId );
      return item.id.playlistId;
    }

    channelIds.push( item.id.channelId );
    return item.id.channelId;
  });

  const videos: Array<Video> = await getVideoById( videoIds.join( ',' ) );
  const playlists: Array<Playlist> = await getPlaylistById( playlistIds.join( ',' ) );
  const channels: Array<Channel> = await  getChannelById( channelIds.join( ',' ) );

  // ids array needed to save order of contents

  return ids.map( item => {
    if ( videos[0] && videos[0].id === item ) return videos.shift();
    if ( playlists[0] && playlists[0].id === item ) return playlists.shift();
    return channels.shift();
  });
}