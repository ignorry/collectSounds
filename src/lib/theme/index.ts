import { DefaultTheme } from "styled-components";
import { ThemeEnum } from "../../models/theme";
import { baseTheme } from "./baseTheme";

export const lightTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.light,

  colors: {
    ...baseTheme.colors,
    main: '#00bb88',
    bg: '#ffffff',
    bgSecondary: '#d8d8d8',
    font: '#000000',
    darkBg: '#d9d9d9'
  },
}

export const darkTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.dark,

  colors: {
    ...baseTheme.colors,
    bg: '#000000',
    bgSecondary: '#2a2a2a',
    font: '#ffffff',
    darkBg: '#121212'
  },
}

/**
 * get theme depending on isDark parameter
 * @param {boolean} isDark 
 * @returns {DefaultTheme}
 */
export const getTheme = ( isDark: boolean ): DefaultTheme => isDark ? darkTheme : lightTheme;