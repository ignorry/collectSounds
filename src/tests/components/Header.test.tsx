import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { getTheme } from "../../lib/theme";
import GlobalStyles from "../../lib/styled/global";
import { BrowserRouter } from "react-router-dom";

import Header from "../../components/primitives/Header";

describe( 'Back component', () => {
  it( 'display picture ', () => {
    const result = render(
      <BrowserRouter>
        <ThemeProvider theme={ getTheme( true ) } >
          <Header />
          <GlobalStyles/>
        </ThemeProvider>
      </BrowserRouter>
    );

    const pic = result.container.querySelector( 'svg' );

    expect( pic ).toBeDefined();
  });

  it( 'calls callback on click', () => {
    const callback = jest.fn();

    const result = render(
      <BrowserRouter>
        <ThemeProvider theme={ getTheme( true ) } >
          <Header callback={callback} />
          <GlobalStyles/>
        </ThemeProvider>
      </BrowserRouter>
    );

    result.container.querySelector( 'button' ).click();

    expect( callback ).toBeCalled();
  });

  it( 'matches to snapshot', () => {
    const result = render(
      <BrowserRouter>
        <ThemeProvider theme={ getTheme( true ) } >
          <Header />
          <GlobalStyles/>
        </ThemeProvider>
      </BrowserRouter>
    );

    expect( result ).toMatchSnapshot();
  });
});