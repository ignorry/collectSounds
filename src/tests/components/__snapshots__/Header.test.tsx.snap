// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Back component matches to snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hKMtZM fGVoVJ"
      >
        <div
          class="sc-jSMfEi gOnElc"
        >
          <button
            class="sc-dkzDqf fcjXBB"
          >
            <svg
              class="sc-bczRLJ bygNPW"
            >
              <use
                href="http://localhost/icons.svg#back"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="sc-eCYdqJ jCVaEc"
      >
        <div
          class="sc-jSMfEi gOnElc"
        >
          <button
            class="sc-dkzDqf fcjXBB"
          >
            <svg
              class="sc-bczRLJ bygNPW"
            >
              <use
                href="http://localhost/icons.svg#back"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hKMtZM fGVoVJ"
    >
      <div
        class="sc-jSMfEi gOnElc"
      >
        <button
          class="sc-dkzDqf fcjXBB"
        >
          <svg
            class="sc-bczRLJ bygNPW"
          >
            <use
              href="http://localhost/icons.svg#back"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="sc-eCYdqJ jCVaEc"
    >
      <div
        class="sc-jSMfEi gOnElc"
      >
        <button
          class="sc-dkzDqf fcjXBB"
        >
          <svg
            class="sc-bczRLJ bygNPW"
          >
            <use
              href="http://localhost/icons.svg#back"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
